Expecta/NSArray+Difference.m

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2020 Lightricks. All rights reserved.
+// Created by Maxim Grabarnik.
+
+#import "NSArray+Difference.h"
+
+#import <assert.h>
+
+#import "ExpectaSupport.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+API_AVAILABLE(ios(13.0))
+typedef NSOrderedCollectionChange<NSObject *> Change;
+
+API_AVAILABLE(ios(13.0))
+@interface ChangePair : NSObject
+
+- (instancetype)init NS_UNAVAILABLE;
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal;
+
+@property (nonatomic, readonly) Change *insertion;
+
+@property (nonatomic, readonly) Change *removal;
+
+@end
+
+@implementation ChangePair
+
+- (instancetype)initWithInsertion:(Change *)insertion removal:(Change *)removal {
+  if (self = [super init]) {
+    _insertion = insertion;
+    _removal = removal;
+  }
+  return self;
+}
+
+@end
+
+@implementation NSArray (Difference)
+
+- (nonnull NSArray<NSString *> *)differenceFrom:(id)object
+    API_AVAILABLE(ios(13.0)) {
+  assert([object isKindOfClass:NSArray.class]);
+  NSArray *expected = self;
+  NSArray *actual = (NSArray *)object;
+
+  NSOrderedCollectionDifference<NSObject *> *diff = [expected differenceFromArray:actual withOptions:NSOrderedCollectionDifferenceCalculationInferMoves];
```

Comment:
can you explain how this flag helps?

Reply:
with this flag the api tries to correlate removals and insertions to find reorders. 
Example: diff between [1,2,3] and [1,3,2] with this flag will say that value 3 is at the wrong index.
Otherwise it's a bunch of removals and insertions that are harder to read.

Reply:
Yes, but how do you know it's a move or just insertion?

Reply:
The docs says the moves are relevant when using identity instead of equality. This kinda makes sense since if the original array is [1,2,3] and the new array is [2,1,1] it's hard to define moves. What happens in that case?

Reply:
That's a good point. Unrelated to moves the question with no clear answer to (in your example) is "which of the ones is the extra one?". 
I don't know how they determine it internally. Judging from result below, distance in terms of index is probably a factor. The result seems reasonable.
What documentation do you refer to ? NSArray methods are pretty undocumented.

`expect(@[@2, @1, @1]).to.equal(@[@1, @2, @3])`:
```
Mismatch at index (2)
	Expected: (3)
	Actual: (1)
Wrong index for value: (2)
	Expected at index (1)
	Actual index is (0)
```

without inferring moves option the difference looses clarity imo:
```
Mismatch at index (2)
	Expected: (3)
	Actual: (1)
Actual contains extra item
	index: (0)
	value: (2)
Actual misses an item at index (1)
	Expected: (2)
```

Reply:
the header files contain documentation. So the `associateIndex` is the key here? I hope this works for more complex values.
IMO both are a bit hard to understand, but that is another problem. Imagine if instead `Expected: (3)` You'd see `Expected <very long description>`.

Reply:
No doubt the more complex values are the harder it is to understand. Hopefully in majority of mismatch cases user would have some idea about the values, which will ease up understanding.
Also in case of differentiable values mismatched items would be reduced to the expected/actual diff. (see lines 74-78 in NSArray+Difference.m)
Let's release it into the wild and interrogate people afterwards.

By the way, swift counterpart starts from  "Different count: Expected 17, Actual 24" when the number of items is different. We can add it here as well. WDYT?

Reply:
The count is definitely important.